file_formats/cmt_format.ml

```diff
@@ -197,12 +213,169 @@ let clear_env binary_annots =
 
   else binary_annots
 
+let iter_on_usages ~index =
+  let f ~namespace env path lid =
+    let not_ghost { Location.loc = { loc_ghost; _ }; _ } = not loc_ghost in
+    if not_ghost lid then
+      try
+        let shape = Env.shape_of_path ~namespace env path in
+        let shape = Local_reduce.weak_reduce env shape in
+        if not (Shape.is_closed shape) then
+          index := (Unresolved shape, lid) :: !index
+        else Option.iter
+          (fun uid -> index := (Resolved uid, lid) :: !index)
+          shape.Shape.uid
+      with Not_found -> ()
```

Comment:
If `Local_reduce` doesn't raise, then I'd prefer using `match Env.shape_of_path ... with exception ...`.

Reply:
I renamed the one in Shapes and added more information on its usage's restrictions in 1d5b0a3

Reply:
Indeed, `Local_reduce` should not raise™️. I made the change in 2551bf7bfb.